matrix/src/lib.rs

```diff
@@ -52,25 +44,38 @@ impl Display for Dimensions {
     }
 }
 
-/// A `Matrix` that supports randomly accessing particular coefficients.
-pub trait MatrixGet<T>: Matrix<T> {
-    fn get(&self, r: usize, c: usize) -> T;
-}
+pub trait Matrix<T: Send + Sync>: Send + Sync {
+    fn width(&self) -> usize;
+    fn height(&self) -> usize;
+
+    fn dimensions(&self) -> Dimensions {
+        Dimensions {
+            width: self.width(),
+            height: self.height(),
+        }
+    }
+
+    fn get(&self, r: usize, c: usize) -> T {
+        self.row(r).nth(c).unwrap()
+    }
 
-/// A `Matrix` that supports randomly accessing particular rows.
-pub trait MatrixRows<T>: Matrix<T> {
-    type Row<'a>: IntoIterator<Item = T>
+    type Row<'a>: Iterator<Item = T> + Send + Sync
     where
         Self: 'a;
 
     fn row(&self, r: usize) -> Self::Row<'_>;
 
     fn rows(&self) -> impl Iterator<Item = Self::Row<'_>> {
-        (0..self.height()).map(|r| self.row(r))
+        (0..self.height()).map(move |r| self.row(r))
     }
 
-    fn row_vec(&self, r: usize) -> Vec<T> {
-        self.row(r).into_iter().collect()
+    fn par_rows(&self) -> impl IndexedParallelIterator<Item = Self::Row<'_>> {
+        (0..self.height()).into_par_iter().map(move |r| self.row(r))
+    }
+
+    // Opaque return type implicitly captures &'_ self
+    fn row_slice(&self, r: usize) -> impl Deref<Target = [T]> {
```

Comment:
Interesting, hadn't thought of doing it this way... I think it makes AIRs a tad less nice but maybe worth it for the more unified API

Reply:
Yeah, I changed from `AsRef<[T]>` to `Deref<Target = [T]>` so you can just do `local[0]` instead of `local.as_ref()[0]`. Still sometimes need to fiddle with it though, like having to do `(*local).borrow()` for the column casting